### ISSUE TITLE
google_cloudfunctions_function: patch egress setting 400 bad request

### ISSUE THREAD

**Issue by BartBucknill** [hashicorp/terraform-provider-google#8719]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version
v0.14.4

### Affected Resource(s)

* google_cloudfunctions_function

### Terraform Configuration Files

1. Create function
```tf
resource "google_cloudfunctions_function" "my_func" {
  count   = local.deploy_cloud_function
  name    = "fi_${local.namespace}_my-func"
  runtime = "go111"
  region  = local.app_engine_region

  available_memory_mb   = 256
  source_archive_bucket = "cloud-functions"
  source_archive_object = "my-func/${local.cloud_function_version["my_func"]}.zip"

  event_trigger {
    event_type = "google.pubsub.topic.publish"
    resource   = google_pubsub_topic.integrations["my-func"].name
  }

  timeout     = 540
  entry_point = "MyFunc"

  service_account_email = "my-account@${local.project}.iam.gserviceaccount.com"

  vpc_connector                 = google_vpc_access_connector.my_vpc_connector.self_link
}
```

2. Patch Function with `vpc_connector_egress_settings`
```tf
resource "google_cloudfunctions_function" "my_func" {
  count   = local.deploy_cloud_function
  name    = "fi_${local.namespace}_my-func"
  runtime = "go111"
  region  = local.app_engine_region

  available_memory_mb   = 256
  source_archive_bucket = "cloud-functions"
  source_archive_object = "my-func/${local.cloud_function_version["my_func"]}.zip"

  event_trigger {
    event_type = "google.pubsub.topic.publish"
    resource   = google_pubsub_topic.integrations["my-func"].name
  }

  timeout     = 540
  entry_point = "MyFunc"

  service_account_email = "my-account@${local.project}.iam.gserviceaccount.com"

  vpc_connector                 = google_vpc_access_connector.my_vpc_connector.self_link
  vpc_connector_egress_settings = "ALL_TRAFFIC"
}
```

### Debug Output

Selection from debug logs https://gist.github.com/BartBucknill/72ff522c9686021d13ce4e06fcc534e1

### Expected Behavior

Cloud function should have been updated to have VPC connector egress setting `ALL_TRAFFIC`.

### Actual Behavior

Google error 400 bad request.

### Steps to Reproduce

1. `terraform apply` configuration like the first in Terraform Configuration Files above, e.g. _without_ `vpc_connector_egress_settings`
2. `terraform apply` configuration like the second in Terraform Configuration Files above, e.g. _with_ `vpc_connector_egress_settings = "ALL_TRAFFIC"`

### Important Factoids

* This does not occur when **creating** the function, only when **updating** an existent function. As such a **workaround** is possible by first running `taint` on the function, and then `apply`.
* A 400 error also occurs if a function is already configured to have VPC connector setting `ALL_TRAFFIC`, and an update is made to change it to `PRIVATE_RANGES_ONLY`.
EDIT
* Update succeeds if both fields `vpc_connector` and `vpc_connector_egress_settings` are added at the same time

### References

* https://github.com/hashicorp/terraform-provider-google/issues/6210

**Comment by edwardmedia** [comment]
@BartBucknill looks like the api only accepts both together (like what you have found) even you just try to update one.  We could update the provider to support that.

```
PATCH /v1/projects/myproject/locations/us-central1/functions/function-test?alt=json&prettyPrint=false&updateMask=vpcConnector%2CvpcConnectorEgressSettings HTTP/1.1
Host: cloudfunctions.googleapis.com
{
 "name": "projects/myproject/locations/us-central1/functions/function-test",
 "vpcConnector": "projects/myproject/locations/us-central1/connectors/vpcconn",
 "vpcConnectorEgressSettings": "ALL_TRAFFIC"
}
```

**Comment by edwardmedia** [comment]
@BartBucknill out of curiosity, during my test, did you remove any other attributes in the request payload in your attached log?

**Comment by BartBucknill** [comment]
@edwardmedia Nope, the attached logs are a straight copy of a section of the debug logs, with just the names of resources modified.

**Comment by edwardmedia** [comment]
@BartBucknill from your log, the provider version is v3.45.0. There was a fix in the provider [v3.47.0 ](https://github.com/hashicorp/terraform-provider-google/releases/tag/v3.47.0) which might have resolved the issue. Can you try the latest version to see if you are still able to repro it?

**Comment by BartBucknill** [comment]
@edwardmedia Unfortunately with google provider at `v3.47.0`, I get
```
Error code 13, message: Failed to retrieve function source code
```
Turned on TF logs and manually validated that the source code zip in field `sourceArchiveUrl` in the request, does exist, and that after reverting to `v3.45.0` and forcing replacement by `taint` this error does not occur.

Interestingly, with `v3.47.0` the plan indicates the function will be `updated in-place` with just the change to the `vpc_connector_egress_settings`, but unlike with `v3.45.0` the PATCH request includes much more data than just this field (looks like it may be the full specification of the function), so I can see how this might address the issue.

**Comment by edwardmedia** [comment]
@BartBucknill I don't know what you meant below. The error is now on `sourceArchiveUrl`. I don't see any difference on this field below versions. Without seeing the specific value for this field, it is hard to guess.  Can you import this resource with the **latest** version of the provider and compare it with the existing state (v3.45.0)? It might tell you what causes this new error. 

As you have found out, the `PATCH request includes much more data` should solve the original problem you encountered.

>Turned on TF logs and manually validated that the source code zip in field sourceArchiveUrl in the PATCH request does exist,

**Comment by BartBucknill** [comment]
@edwardmedia Not sure what caused the error on finding source code with v3.47.0, but after I removed and recreated the function, then attempted to update `vpc_connector_egress_settings` again, it worked.
Happy for this issue to be closed as it has been fixed in v3.47.0.
Thanks very much for your help @edwardmedia 😃
